src/linter/root.go

```diff
@@ -1111,6 +1134,40 @@ func (d *RootWalker) enterConstList(lst *stmt.ConstList) bool {
 	return false
 }
 
+func (d *RootWalker) leaveFile() {
+	d.checkFilename()
+}
+
+func (d *RootWalker) checkFilename() {
+	fileTypes := len(d.meta.Classes) + len(d.meta.Traits)
+	if fileTypes == 0 {
+		return
+	}
+
+	// If there are any types, at least one of them
+	// should match the fileBaseName.
+	badFilename := true
+	// We're checking "base" part of the class name with HasSuffix.
+	// Leading "\" ensures that we don't match only part of the base name.
+	// Note that "meta.Classes" includes interfaces as well.
+	fileBaseName := `\` + strings.TrimSuffix(filepath.Base(d.filename), filepath.Ext(d.filename))
+	for classFullname := range d.meta.Classes {
+		if strings.HasSuffix(classFullname, fileBaseName) {
+			badFilename = false
+		}
+	}
+	for traitFullname := range d.meta.Traits {
+		if strings.HasSuffix(traitFullname, fileBaseName) {
+			badFilename = false
+		}
+	}
+
+	if badFilename {
+		// TODO: which node to use to report file-level issues?
```

Comment:
I suppose that we can use any class node or trait node. But we should save them somewhere then :).